File: src/JsonRpcClient/JsonRpcClient.fs
namespace TcpEcho

open System
open System.Buffers
open System.Net.Sockets
open System.IO.Pipelines
open System.Text
open System.Threading.Tasks

type CommunicationUnsuccessfulException(msg: string, innerException: Exception) =
    inherit Exception(msg, innerException)

exception IncompleteResponseException of string
type TimeoutOrResult<'a> = Timeout | Result of 'a

[<AbstractClass>]
type Client(endpoint: string, port: int) =
    let minimumBufferSize = 1024

    let withTimeout (timeout : int) (xAsync: Async<'a>) = async {
        let read = async {
            let! value = xAsync
            return value |> Result |> Some
        }
        let delay = async {
            do! Async.Sleep(timeout)
            return Some Timeout
        }

        let! result = Async.Choice([read; delay])
        match result with
        | Some x -> return x
        | None -> return Timeout
    }

    let rec writeToPipeAsync (writer: PipeWriter) (socket: Socket) = async {
        try
            let segment = Array.zeroCreate<byte> minimumBufferSize |> ArraySegment
            let! read = socket.ReceiveAsync(segment, SocketFlags.None) |> Async.AwaitTask |> withTimeout socket.ReceiveTimeout

            match read with
            | Timeout ->
                return writer.Complete(TimeoutException("Socket read timed out"))
            | Result 0 ->
                return writer.Complete()
            | Result bytesRead ->
                segment.Array.CopyTo(writer.GetMemory(bytesRead))
                writer.Advance bytesRead
                let! flusher = writer.FlushAsync().AsTask() |> Async.AwaitTask
                if flusher.IsCompleted then
                    return writer.Complete()
                else
                    return! writeToPipeAsync writer socket
        with ex -> return writer.Complete(ex)
    }

    let rec readFromPipeAsync (reader: PipeReader) (sb: StringBuilder) = async {
        try
            let! result = reader.ReadAsync().AsTask() |> Async.AwaitTask

            let buffer: ReadOnlySequence<byte> = result.Buffer

            buffer |> ref |> BuffersExtensions.ToArray |> Encoding.UTF8.GetString |> sb.Append |> ignore

            reader.AdvanceTo(buffer.End)

            match result.IsCompleted with
            | true ->
                reader.Complete()
                return sb.ToString()
            | false ->
                return! readFromPipeAsync reader sb
        with
            // If we got an TimeoutException anywhere in the exception chain
            | :? AggregateException as ae when ae.Flatten().InnerExceptions |> Seq.exists (fun x -> x :? TimeoutException) ->
                return raise <| IncompleteResponseException (sb.ToString())
    }

    let CallImplAsync (json: string) =
        async {
            use socket = new Socket(SocketType.Stream, ProtocolType.Tcp)
            socket.ReceiveTimeout <- 500
            socket.SendTimeout <- 500
            do! socket.ConnectAsync(endpoint, port) |> Async.AwaitTask

            let segment = UTF8Encoding.UTF8.GetBytes(json + Environment.NewLine) |> ArraySegment

            let! send = socket.SendAsync(segment, SocketFlags.None) |> Async.AwaitTask |> withTimeout socket.SendTimeout
            match send with
            | Result _ ->
                let pipe = Pipe()

                let! _ = writeToPipeAsync pipe.Writer socket |> Async.StartChild
                return! readFromPipeAsync pipe.Reader (StringBuilder())
            | Timeout -> return raise (CommunicationUnsuccessfulException("Socket send timed out", null))
        }

    abstract member CallAsync: string -> Async<string>
    abstract member CallAsyncAsTask: string -> Task<string>

    default __.CallAsync (json: string) =
        async {
            try
                return! CallImplAsync json
            with
            | IncompleteResponseException response as ex when String.IsNullOrWhiteSpace(response) ->
                return raise <| CommunicationUnsuccessfulException("Empty response from socket", ex)
            | IncompleteResponseException _ as ex ->
                return raise <| ex
            | :? AggregateException as ae when ae.Flatten().InnerExceptions |> Seq.exists (fun x -> x :? SocketException) ->
                return raise <| CommunicationUnsuccessfulException(ae.Message, ae)
            | :? SocketException as ex ->
                return raise <| CommunicationUnsuccessfulException(ex.Message, ex)
        }

    default this.CallAsyncAsTask (json: string) =
        this.CallAsync json |> Async.StartAsTask

File: src/JsonRpcClient/JsonRpcClient.cs

using System;
using System.Buffers;
using System.IO.Pipelines;
using System.Net.Sockets;
using System.Runtime.InteropServices;
using System.Text;
using System.Threading.Tasks;


namespace TcpEcho {

    public class CommunicationUnsuccessfulException : Exception
    {
        internal CommunicationUnsuccessfulException(string msg, Exception innerException) : base(msg, innerException)
        {
        }
    }

    public abstract class JsonRpcClient {
        private const int minimumBufferSize = 1024;
        private readonly TimeSpan tcpTimeout = TimeSpan.FromSeconds(1);
        private readonly string endpoint;
        private readonly int port;

        public JsonRpcClient(string endpoint, int port) {
            this.endpoint = endpoint;
            this.port = port;
        }

        private async Task<string> CallImpl (string json) {
            using (Socket socket = new Socket (SocketType.Stream, ProtocolType.Tcp)) {
                socket.ReceiveTimeout = (int)tcpTimeout.TotalMilliseconds;

                var connectTimeOut = Task.Delay (tcpTimeout);
                var completedConnTask = await Task.WhenAny (connectTimeOut, socket.ConnectAsync (endpoint, port));
                if (completedConnTask == connectTimeOut) {
                    throw new TimeoutException("connect timed out");
                }

                byte[] bytesToSend = Encoding.UTF8.GetBytes (json + Environment.NewLine);
                socket.Send (bytesToSend);

                var pipe = new Pipe ();
                var writing = WriteToPipeAsync (socket, pipe.Writer);
                var reading = ReadFromPipeAsync (pipe.Reader);

                var readAndWriteTask = Task.WhenAll (reading, writing);
                var readTimeOut = Task.Delay (tcpTimeout);
                var completedReadTask = await Task.WhenAny (readTimeOut, readAndWriteTask);
                if (completedReadTask == readTimeOut) {
                    throw new TimeoutException("reading/writing from socket timed out");
                }

                return await reading;
            }
        }

        protected async Task<string> Call(string json)
        {
            try
            {
                var reply = await CallImpl(json);
                if (String.IsNullOrEmpty(reply)) {
                    throw new Exception("Some problem while reading reply: it was empty");
                }
                return reply;
            }
            catch (SocketException ex)
            {
                throw new CommunicationUnsuccessfulException(ex.Message, ex);
            }
            catch (TimeoutException ex)
            {
                throw new CommunicationUnsuccessfulException(ex.Message, ex);
            }
        }

        private static async Task WriteToPipeAsync (Socket socket, PipeWriter writer) {
            await Task.Yield();
            while (true)
            {
                if (socket.Available > 0)
                {
                    var memory = writer.GetMemory(minimumBufferSize);
                    var bytesRead = await socket.ReceiveAsync(memory, SocketFlags.None);
                    writer.Advance(bytesRead);
                }

                var flushResult = await writer.FlushAsync ();
                if (flushResult.IsCompleted) {
                    break;
                }
            }

            writer.Complete ();
        }

        private static async Task<string> ReadFromPipeAsync (PipeReader reader) {
            var strResult = String.Empty;
            while (true) {
                ReadResult result = await reader.ReadAsync ();
                ReadOnlySequence<byte> buffer = result.Buffer;

                var content = UTF8String (buffer);
                strResult += content;
                reader.AdvanceTo (buffer.End, buffer.End);

                if (result.IsCompleted || strResult.EndsWith("\n")) {
                    break;
                }
            }

            reader.Complete();
            return strResult;
        }

        private static string UTF8String (ReadOnlySequence<byte> buffer) {
            var result = String.Empty;
            foreach (ReadOnlyMemory<byte> segment in buffer)
            {
#if NETCOREAPP2_1
                result += Encoding.UTF8.GetString (segment.Span);
#else
                result += Encoding.UTF8.GetString (segment);
#endif
            }
            return result;
        }
    }
}


#if NET461

internal static class Extensions
{
    public static Task<int> ReceiveAsync(this Socket socket, Memory<byte> memory, SocketFlags socketFlags)
    {
        var arraySegment = GetArray(memory);
        return SocketTaskExtensions.ReceiveAsync(socket, arraySegment, socketFlags);
    }

    public static string GetString(this Encoding encoding, ReadOnlyMemory<byte> memory)
    {
        var arraySegment = GetArray(memory);
        return encoding.GetString(arraySegment.Array, arraySegment.Offset, arraySegment.Count);
    }

    private static ArraySegment<byte> GetArray(Memory<byte> memory)
    {
        return GetArray((ReadOnlyMemory<byte>)memory);
    }

    private static ArraySegment<byte> GetArray(ReadOnlyMemory<byte> memory)
    {
        if (!MemoryMarshal.TryGetArray(memory, out var result))
        {
            throw new InvalidOperationException("Buffer backed by array was expected");
        }

        return result;
    }
}

#endif
